### PR TITLE
Packet transfer

### PR DESCRIPTION
**Pull request by hamelin**
+ unit tests and integration tests.

### CODE REVIEW

File: itsim/machine/node.py [hunk 1]
```diff
@@ -118,14 +127,16 @@ def _get_port_ephemeral(self) -> Port:
                 break
         raise EphemeralPortsAllInUse()
 
-    def bind(self, pr: PortRepr = 0) -> Socket:
+    def bind(self, protocol: Protocol = Protocol.NONE, pr: PortRepr = 0, as_pid: int = -1) -> Socket:
         """
         Reserves networking resources, in particular a port, for a calling process. If no port is provided, or port 0,
         then a random free port is thus bound. The binding is embedded in a :py:class:`Socket` instance which may be
         then used to send and receive packets of information.
```
Comment (mattcraddock) [inline itsim/machine/node.py]: Should there be a param comment for `protocol` too?

File: itsim/machine/socket.py [hunk 2]
```diff
@@ -114,12 +129,13 @@ def recv(self, timeout: Optional[float] = None) -> Packet:
             raise ValueError("Socket is closed")
 
         try:
-            self._packet_signal.wait(timeout)
+            greensim.select(self._packet_signal, self._close_signal, timeout=timeout)
         except greensim.Timeout:
             raise Timeout()
 
-        if self.is_closed:
+        if self.is_closed or not self._packet_signal.is_on:
```
Comment (mattcraddock) [inline itsim/machine/socket.py]: What case does `not self._packet_signal.is_on` handle?

File: itsim/network/link.py [hunk 3]
```diff
@@ -49,11 +61,29 @@ def iter_nodes(self) -> Iterator[_Node]:
...
+        else:
+            for node, interface in [(node, interface) for node in self.iter_nodes() for interface in node.interfaces()]:
+                if interface.address == hop:
+                    recipients = iter([node])
+                    break
+            else:
+                raise NoSuchAddress(hop)
+
+        packet_latency = next(self._latency)
+        packet_bandwidth = next(self._bandwidth)  # Modeler's responsibility never to provide 0 bandwidth.
```
Comment (mattcraddock) [inline itsim/network/link.py]: What does this comment mean? Why not do validation?

File: tests/integ/test_packet_transfer.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+        try:
+            socket.recv(5 * S)
+            pytest.fail()
+        except Timeout:
+            assert now() > 5.0
+
+
+def server(thread: Thread) -> None:
+    with thread.process.node.bind(Protocol.UDP, 9887) as socket:
+        while True:
```
Comment (mattcraddock) [inline tests/integ/test_packet_transfer.py]: Why is this necessary? It looks like only one packet is being transferred

File: tests/integ/test_packet_transfer.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+def test_packet_transfer():
+    sim = Simulator()
+
+    link = Link("10.11.12.0/24", latency=uniform(100 * MS, 200 * MS), bandwidth=constant(100 * MbPS))
+    pinger = Endpoint().connected_to(link, "10.11.12.10")
+    pinger.fork_exec_in(sim, 0.1, client)
+    ponger = Endpoint().connected_to(link, "10.11.12.20")
+    ponger.fork_exec(sim, server)
+
+    sim.run(10.0 * S)
```
Comment (mattcraddock) [inline tests/integ/test_packet_transfer.py]: Since this only triggers the `assert` statements indirectly, I think it is best to have them alter a global variable or throw an ad hoc exception so that there is evidence that both `client` and `server` actually ran

File: tests/machine/test_node.py [hunk 6]
```diff
@@ -255,3 +255,91 @@ def test_recv_socket_timeout_none(socket80):
...
+
+
+def test_send_packet(endpoint_2links, link_small, link_large):
+    for dest, link, relay in [
+        ("192.168.1.67", link_small, "192.168.1.67"),
+        ("10.10.192.245", link_large, "10.10.192.245"),
+        ("172.92.0.2", link_large, "10.10.128.1")
+    ]:
+        for interface, route in (
+            (interface, route) for interface in endpoint_2links.interfaces() for route in interface.forwardings
```
Comment (mattcraddock) [inline tests/machine/test_node.py]: Is this something you could do with zip? https://docs.python.org/3.5/library/functions.html#zip

File: tests/machine/test_node.py [hunk 7]
```diff
@@ -255,3 +255,91 @@ def test_recv_socket_timeout_none(socket80):
...
+        ):
+            if as_address(dest) in route.cidr:
+                source = interface.address
+                break
+        else:
+            pytest.fail()
```
Comment (mattcraddock) [inline tests/machine/test_node.py]: What is this testing? Why not set up the values directly?

File: itsim/machine/node.py (hunk 1 above)
Comment (hamelin) [reply itsim/machine/node.py]: Absolutely!

Comment (hamelin) [reply itsim/machine/node.py]: Fixed in c905a6d

File: itsim/machine/socket.py (hunk 2 above)
Comment (hamelin) [reply itsim/machine/socket.py]: This is likely an accident -- #34 was meant to document this modification.

In any case, this pertains to the `greensim.select` statement above. If it returns without raising, then either a packet has been received, or the socket has been signaled as closed -- these alternatives being mutually exclusive. I was wary that the `is_closed` status variable and `_close_signal` may be out of sync, and assumed that if the socket was closed, then it was unlikely that the packet signal would be turned on...

Now that you force me to write down the logic, I see how convoluted it actually is. I will prepare a patch to simplify all this.

File: itsim/network/link.py (hunk 3 above)
Comment (hamelin) [reply itsim/network/link.py]: Fixed in 1fcee54

File: tests/integ/test_packet_transfer.py (hunk 4 above)
Comment (hamelin) [reply tests/integ/test_packet_transfer.py]: It demonstrates the implementation of a typical server program, giving this test documentation quality. Should we follow the principle that even integration tests should be the simplest that can evidence the behaviour we mean to observe and enforce? This is the main reason I would see behind removing this loop.

File: tests/integ/test_packet_transfer.py (hunk 5 above)
Comment (hamelin) [reply tests/integ/test_packet_transfer.py]: Ah *now* I understand why your own integration tests carry so many global variables. Clever!

Comment (hamelin) [reply tests/integ/test_packet_transfer.py]: Fixed in ed6901a

File: tests/machine/test_node.py (hunk 6 above)
Comment (hamelin) [reply tests/machine/test_node.py]: No. `zip` pairs two sequences element-to-element. This double-for list comprehension embodies nested loops, bringing them into a single loop that can be `break`ed out of.

File: tests/machine/test_node.py (hunk 7 above)
Comment (hamelin) [reply tests/machine/test_node.py]: The else clause to a for loop runs if the loop is not `break`ed out of. In this case, it happens if no suitable `source` can be found. This is a mistake and it should be brought up. I will add a message to the `fail` call to make it clearer.

Comment (hamelin) [reply tests/machine/test_node.py]: Fixed in 341e3b2

File: itsim/network/link.py (hunk 3 above)
Comment (mattcraddock) [reply itsim/network/link.py]: Would `bounded` work? https://github.com/ElementAI/greensim/blob/master/greensim/random.py#L44

File: tests/machine/test_node.py (hunk 6 above)
Comment (hamelin) [reply tests/machine/test_node.py]: Test simplified in 5ce90fd.

File: tests/integ/test_packet_transfer.py (hunk 4 above)
Comment (hamelin) [reply tests/integ/test_packet_transfer.py]: Test simplified in 775404f

File: tests/integ/test_broadcast.py [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+
+    def server(self, thread: Thread):
+        with thread.process.node.bind(Protocol.UDP, 10000) as socket:
+            while True:
+                packet = socket.recv()
+                self._peers.add(packet.source.hostname_as_address())
+
+    def client(self, thread: Thread):
+        with thread.process.node.bind(Protocol.UDP) as socket:
+            while True:
```
[the remark this answers is not in the document]
Comment (hamelin) [reply tests/integ/test_broadcast.py]: Agreed. This was also an instance of demonstrating usage. Will not do again.
